[FEATURE] Add Anthropic LLM support via `anthropic` Python SDK

dot-agi opened the pull request:
## 📥 Pull Request

**📘 Description**
This PR adds functions in the `llm_tracker.py` file for the asynchronous and synchronous Anthropic SDK support. An example notebook `anthropic_example` under `examples/anthropic` is also added to demo the integration.

**🎯 Goal**
To extend the AgentOps capability to the Anthropic LLMs.

**🔍 Additional Context**
Tool Calling in Anthropic Python SDK seems flaky and might not result in correct responses,

**🧪 Testing**
Tested using the provided example notebook.

Thank you for your contribution to Agentops!

dot-agi committed:
fix typo

dot-agi committed:
add anthropic support

dot-agi committed:
add example file

dot-agi committed:
fixed linting using black

dot-agi committed:
add time travel support for anthropic

dot-agi committed:
linting

dot-agi committed:
fix kwargs key to get prompt

dot-agi committed:
fix for extracting tokens from Message.usage

dot-agi committed:
minor fix for output tokens

dot-agi committed:
remove anthropic example python file

dot-agi committed:
fix completions not show in session

dot-agi committed:
some more fixes and cleanup

dot-agi committed:
add Message object to pydantic models

dot-agi committed:
fix typo

dot-agi committed:
overhaul anthropic code

dot-agi committed:
linting

dot-agi committed:
add anthropic example notebook

dot-agi committed:
linting

dot-agi commented:
@bboynton97 @areibman This is ready for review!

areibman commented on examples/anthropic/anthropic_example.ipynb:
```diff
@@ -0,0 +1,175 @@
+    "    elif event.type == \"message_stop\":\n",
+    "        print(\"\\n\")\n",
+    "        print(response)\n",
+    "        print(\"\\n\")"
```
The second example doesn't print anything for some reason

<img width="613" alt="image" src="https://github.com/user-attachments/assets/d3d7a37c-f12b-452b-bdbf-fab4a8938728">

areibman replied:
However, it does get recorded in AgentOps:
<img width="1450" alt="image" src="https://github.com/user-attachments/assets/7376a6ef-5503-4d63-9989-6bc03b81a1a6">

dot-agi replied:
I re-ran the code and found out the tool calling isn't doing what it's expected, in this case, performing a web search). This is why it's only printing the "thinking" part but not the "answer". I have 2 solutions in mind for a fix -
* Precisely describe the web search in the tool
* Use the [anthropic-tools](https://github.com/anthropics/anthropic-tools) package which is in alpha stage

I will be testing both of these and push a fix soon.

dot-agi replied:
I tested with both of these and so far no luck since there is no innate web search support for the models. However, another solution would be to use the anthropic-tools package with a custom Python class that uses another package to perform web searches, getting us the results we want. Or we could skip such questions and ask something else in the prompt.

Please let me know what are your thoughts on this.

areibman commented on agentops/llm_tracker.py:
```diff
@@ -621,6 +624,130 @@ async def async_generator():
+            self.llm_event.completion = {
+                "role": "assistant",
+                "content": response.content,
+            }
```
<img width="492" alt="image" src="https://github.com/user-attachments/assets/4f8b2b54-0666-4913-bd0e-29c715c1f756">

Issue where async completion non-stream doesn't get formatted as an LLMEvent correctly. Sample code:

```
import asyncio
from anthropic import AsyncAnthropic

client = AsyncAnthropic(
    # This is the default and can be omitted
    api_key=os.environ.get("ANTHROPIC_API_KEY"),
)


async def main() -> None:
    message = await client.messages.create(
        max_tokens=1024,
        messages=[
            {
                "role": "user",
                "content": "Hello, Claude",
            }
        ],
        model="claude-3-opus-20240229",
    )
    print(message.content)


await main()```

dot-agi replied:
Thanks for testing this out! I will push a fix soon.

dot-agi replied:
Fixed in the latest commit!

areibman commented on examples/anthropic/anthropic_example.ipynb:
Can you add an example for Async?
```

import asyncio
from anthropic import AsyncAnthropic

client = AsyncAnthropic(
    # This is the default and can be omitted
    api_key=os.environ.get("ANTHROPIC_API_KEY"),
)


async def main() -> None:
    message = await client.messages.create(
        max_tokens=1024,
        messages=[
            {
                "role": "user",
                "content": "Hello, Claude",
            }
        ],
        model="claude-3-opus-20240229",
    )
    print(message.content)


await main()```

dot-agi replied:
Sure!

areibman committed:
added readme examples

areibman reviewed (changes requested):
Really great PR! I left a few comments. Namely, Async isn't parsing data as we'd expect. Otherwise looks good

dot-agi committed:
fix incorrect attribute access for the model content

dot-agi committed:
add async example

areibman commented:
Unfortunate timing for this PR. @bboynton97 is currently splitting llm_tracker.py into multiple files for readability. When that merge goes through, we'll make some changes to make this one work

dot-agi commented:
> Unfortunate timing for this PR. @bboynton97 is currently splitting llm_tracker.py into multiple files for readability. When that merge goes through, we'll make some changes to make this one work

I'd like to help, would be good for me to understand the working of the software in detail and I can use that information to draft a contributing guidelines document and/or observe where tests are needed.

areibman commented:
> > Unfortunate timing for this PR. @bboynton97 is currently splitting llm_tracker.py into multiple files for readability. When that merge goes through, we'll make some changes to make this one work
> 
> I'd like to help, would be good for me to understand the working of the software in detail and I can use that information to draft a contributing guidelines document and/or observe where tests are needed.

This is the PR in question: https://github.com/AgentOps-AI/agentops/pull/346

From your end, this would just mean splitting the Anthropic handlers into its own file. The tricky part is just merging those changes from that branch into yours.

areibman commented:
@HowieG Can you add the Anthropic notebook to the smoke test suite?

dot-agi commented:
> > > Unfortunate timing for this PR. @bboynton97 is currently splitting llm_tracker.py into multiple files for readability. When that merge goes through, we'll make some changes to make this one work
> > 
> > 
> > I'd like to help, would be good for me to understand the working of the software in detail and I can use that information to draft a contributing guidelines document and/or observe where tests are needed.
> 
> This is the PR in question: #346
> 
> From your end, this would just mean splitting the Anthropic handlers into its own file. The tricky part is just merging those changes from that branch into yours.

Cool, I will look into it and find a solution.

HowieG commented:
> @HowieG Can you add the Anthropic notebook to the smoke test suite?

Any notebook added to tests or examples folders will be smoke tested automatically. Except any we've excluded

dot-agi committed:
refactor code

dot-agi commented:
> > > > Unfortunate timing for this PR. @bboynton97 is currently splitting llm_tracker.py into multiple files for readability. When that merge goes through, we'll make some changes to make this one work
> > > 
> > > 
> > > I'd like to help, would be good for me to understand the working of the software in detail and I can use that information to draft a contributing guidelines document and/or observe where tests are needed.
> > 
> > 
> > This is the PR in question: #346
> > From your end, this would just mean splitting the Anthropic handlers into its own file. The tricky part is just merging those changes from that branch into yours.
> 
> Cool, I will look into it and find a solution.

I rebased and refactored the code. I believe that this PR should merge to `llm-handler-refactor` if that's where the changes are right now.

dot-agi committed:
fix function name

dot-agi committed:
linting

dot-agi committed:
Merge branch 'llm-handler-refactor' into anthropic-ops

dot-agi committed:
add provider tests

braelyn-ai commented:
hey @the-praxs! thank you for reworking this, it looks fantastic! i'm going to go ahead and merge this into the `llm-handler-refactor` like you set it up to do and make sure you get credit on the subsequent PR as well :)

thank you for your work here! its very appreciated
